src/path.cr

```diff
@@ -0,0 +1,847 @@
+          str << @type.separator unless str.bytesize == 0
+        end
+
+        last_ended_with_separator = @type.ends_with_separator?(part)
+
+        str.write part.unsafe_byte_slice(byte_start, byte_count)
+      end
+    end
+
+    new_instance new_name
+  end
+
+  # Appends the given *part* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"] / "bar" / "baz"     # => Path["foo/bar/baz"]
+  # Path["foo/"] / Path["/bar/baz"] # => Path["foo/bar/baz"]
+  # ```
+  def /(part : Path | String) : Path
+    join(part)
+  end
+
+  # :nodoc:
+  def <<(part : Path | String) : Path
+    @name = join(part).@name
+    self
+  end
+
+  # Resolves path *name* in this path's parent directory.
+  #
+  # Raises `Path::Error` if `#parent` is `nil`.
+  def sibling(name : Path | String) : Path?
+    if parent = parent()
+      parent.join(name)
+    else
+      raise Error.new("Can't resolve sibling for a path without parent directory")
+    end
+  end
+
+  # Compares this path to *other*.
+  # The comparison is performed strictly lexically: `foo` and `./foo` are *not*
+  # treated as equal. To compare paths semantically, they need to be normalized
+  # and converted to the same flavour.
+  #
+  # Comparison is case-sensitive for POSIX paths and case-insensitive for
+  # Windows paths.
+  def <=>(other : Path)
+    ord = @name.compare(other.@name, case_insensitive: windows?)
+    return ord if ord != 0
+
+    @type <=> other.@type
+  end
+
+  # Returns a path representing the drive component or `nil` if this path does not contain a drive.
+  #
+  # See `#anchor` for the combination of drive and `#root`.
+  #
+  # ```
+  # Path.windows("C:\Program Files").drive    # => Path.windows("C:")
+  # Path.windows("\\host\share\folder").drive # => Path.windows("\\host\share")
+  # ```
+  #
+  # NOTE: Drives are only available for Windows paths. It can either be a drive letter (`C:`) or a UNC share (`\\host\share`).
+  def drive : Path?
+    if drive = @type.drive(@name)
+      new_instance drive
+    end
+  end
+
+  # Returns the root path component of this path or `nil` if it is not rooted.
+  #
+  # See `#anchor` for the combination of `#drive` and root.
+  #
+  # ```
+  # Path.posix("/etc/").root                 # => Path.posix("/")
+  # Path.windows("C:Program Files").root     # => nil
+  # Path.windows("C:\Program Files").root    # => Path.windows("\")
+  # Path.windows("\\host\share\folder").root # => Path.windows("\")
+  # ```
+  def root : Path?
+    if root = @type.root(@name)
+      new_instance root
+    end
+  end
+
+  # Returns the concatenation of `#drive` and `#root`.
+  #
+  # ```
+  # Path.posix("/etc/").anchor                 # => Path.posix("/")
+  # Path.windows("C:Program Files").anchor     # => Path.windows("C:")
+  # Path.windows("C:\Program Files").anchor    # => Path.windows("C:\")
+  # Path.windows("\\host\share\folder").anchor # => Path.windows("\\host\share\")
+  # ```
+  def anchor : Path?
+    root = root()
```

Comment:
Please use `root = self.root` when the method has no argument (more idiomatic), or rename the local variable.
This is valid for the whole pull request.